File: pkg/ecl/proto.go
package ecl

import (
	"sort"
	"strings"

	eclv1 "encr.dev/proto/encore/ecl/v1"
)

// ToProto converts an environment evaluation result into the wire schema in
// encore.ecl.v1. The RuleSet supplies whether each resource kind is managed.
func (rs *RuleSet) ToProto(er *EnvResult) *eclv1.EvaluationResult {
	out := &eclv1.EvaluationResult{}
	for _, r := range er.Results {
		out.Resources = append(out.Resources, rs.resourceToProto(r))
	}
	return out
}

func (rs *RuleSet) resourceToProto(r *Result) *eclv1.Resource {
	res := &eclv1.Resource{
		Kind:    r.Resource.Kind,
		Name:    r.Resource.Name,
		Managed: rs.isManaged(r.Resource.Kind),
	}

	// Scalar properties (reference-valued ones are emitted under references).
	var paths []string
	for path, rp := range r.Properties {
		if rp.Ref == nil {
			paths = append(paths, path)
		}
	}
	sort.Strings(paths)
	for _, path := range paths {
		rp := r.Properties[path]
		res.Properties = append(res.Properties, &eclv1.Property{
			Path:       path,
			Value:      valueToProto(rp.Value),
			Source:     sourceToProto(rp.Source),
			Constraint: mergeConstraints(propConstraints(path, r.Matched)),
		})
	}

	// Reference-valued properties, grouped by path.
	byPath := map[string]*eclv1.Reference{}
	var refPaths []string
	for _, rr := range r.References {
		ref, ok := byPath[rr.Path]
		if !ok {
			ref = &eclv1.Reference{
				Path:       rr.Path,
				TargetKind: rr.TargetKind,
				TargetName: rr.TargetName,
				Unresolved: rr.unresolved,
			}
			byPath[rr.Path] = ref
			refPaths = append(refPaths, rr.Path)
		}
		if rr.Object != nil {
			for _, p := range rr.Object.Props {
				if sv := p.scalar(); sv != nil && sv.Constraint != nil {
					ref.Object = append(ref.Object, &eclv1.PropertyConstraint{
						Path:       p.Path,
						Constraint: mergeConstraints([]Constraint{sv.Constraint}),
					})
				}
			}
		}
	}
	sort.Strings(refPaths)
	for _, p := range refPaths {
		res.References = append(res.References, byPath[p])
	}
	return res
}

// propConstraints collects the scalar constraint expressions for a property
// path from every rule that matched the resource.
func propConstraints(path string, matched []*Rule) []Constraint {
	var cs []Constraint
	for _, r := range matched {
		for _, p := range r.Props {
			if p.Path != path {
				continue
			}
			if sv := p.scalar(); sv != nil && sv.Constraint != nil {
				cs = append(cs, sv.Constraint)
			}
		}
	}
	return cs
}

// mergeConstraints normalizes a set of constraints (all conjoined) into the
// wire Constraint: required flag, min/max bounds, allowed/excluded value sets,
// plus the effective constraint rendered into expr. Returns nil if empty.
func mergeConstraints(cs []Constraint) *eclv1.Constraint {
	if len(cs) == 0 {
		return nil
	}

	var (
		exprParts    []string
		seen         = map[string]bool{}
		required     bool
		lower, upper *Comparison
		exact        *Comparison
		excluded     []Value
		sets         [][]Value
	)
	for _, c := range cs {
		if s := c.String(); !seen[s] {
			seen[s] = true
			exprParts = append(exprParts, s)
		}
		for _, term := range conjuncts(c) {
			switch t := term.(type) {
			case *RequiredConstraint:
				required = true
			case *Comparison:
				switch t.Op {
				case OpEq:
					exact = t
				case OpNeq:
					excluded = append(excluded, t.Value)
				case OpGe, OpGt:
					if lower == nil || tighterLower(t, lower) {
						lower = t
					}
				case OpLe, OpLt:
					if upper == nil || tighterUpper(t, upper) {
						upper = t
					}
				}
			case *OrConstraint:
				if vs, ok := exactAlternatives(t); ok {
					sets = append(sets, vs)
				}
			}
		}
	}

	out := &eclv1.Constraint{
		Required: required,
		Expr:     strings.Join(exprParts, " & "),
	}
	if lower != nil {
		out.Min = &eclv1.Bound{Value: valueToProto(lower.Value), Inclusive: lower.Op == OpGe}
	}
	if upper != nil {
		out.Max = &eclv1.Bound{Value: valueToProto(upper.Value), Inclusive: upper.Op == OpLe}
	}

	var allowed []Value
	switch {
	case exact != nil:
		allowed = []Value{exact.Value}
	case len(sets) > 0:
		allowed = sets[0]
		for _, s := range sets[1:] {
			allowed = intersectValues(allowed, s)
		}
	}
	for _, v := range allowed {
		out.Allowed = append(out.Allowed, valueToProto(v))
	}
	for _, v := range excluded {
		out.Excluded = append(out.Excluded, valueToProto(v))
	}
	return out
}

func intersectValues(a, b []Value) []Value {
	var out []Value
	for _, v := range a {
		if valueInSet(v, b) {
			out = append(out, v)
		}
	}
	return out
}

func valueToProto(v Value) *eclv1.Value {
	out := &eclv1.Value{}
	switch v.Kind {
	case NumberKind:
		out.Kind = &eclv1.Value_NumberValue{NumberValue: v.Num}
	case BoolKind:
		out.Kind = &eclv1.Value_BoolValue{BoolValue: v.Bool}
	case StringKind:
		out.Kind = &eclv1.Value_StringValue{StringValue: v.Str}
	case SizeKind:
		out.Kind = &eclv1.Value_SizeBytes{SizeBytes: v.Num}
		out.Unit = v.unit
	case DurationKind:
		out.Kind = &eclv1.Value_DurationMs{DurationMs: v.Num}
		out.Unit = v.unit
	}
	return out
}

func sourceToProto(s ValueSource) eclv1.ValueSource {
	switch s {
	case SourceExplicit:
		return eclv1.ValueSource_VALUE_SOURCE_EXPLICIT
	case SourceDefault:
		return eclv1.ValueSource_VALUE_SOURCE_DEFAULT
	default:
		return eclv1.ValueSource_VALUE_SOURCE_UNSPECIFIED
	}
}

File: crates/encore_ecl/src/proto.rs
//! Conversion of an ECL evaluation result into the `encore.ecl.v1` wire schema
//! (see `proto/encore/ecl/v1/ecl.proto`).

use std::collections::{HashMap, HashSet};
use std::rc::Rc;

use crate::ast::{CompareOp, Comparison, Constraint, Rule};
use crate::env::EnvResult;
use crate::eval::{EvalResult, RuleSet, ValueSource};
use crate::pb;
use crate::satisfy::{conjuncts, exact_alternatives, tighter_lower, tighter_upper};
use crate::specificity::value_in_set;
use crate::value::{Value, ValueKind};

impl RuleSet {
    /// Converts an environment evaluation result into the `encore.ecl.v1` wire
    /// schema. `self` supplies whether each resource kind is managed.
    pub fn to_proto(&self, er: &EnvResult) -> pb::EvaluationResult {
        pb::EvaluationResult {
            resources: er
                .results
                .iter()
                .map(|r| self.resource_to_proto(r))
                .collect(),
        }
    }

    fn resource_to_proto(&self, r: &EvalResult) -> pb::Resource {
        // Scalar properties (reference-valued ones go under `references`).
        let mut paths: Vec<&String> = r
            .properties
            .iter()
            .filter(|(_, rp)| rp.reference.is_none())
            .map(|(p, _)| p)
            .collect();
        paths.sort();
        let properties: Vec<pb::Property> = paths
            .into_iter()
            .map(|path| {
                let rp = &r.properties[path];
                pb::Property {
                    path: path.clone(),
                    value: Some(value_to_proto(&rp.value)),
                    source: source_to_proto(rp.source),
                    constraint: merge_constraints(&prop_constraints(path, &r.matched)),
                }
            })
            .collect();

        // Reference-valued properties, grouped by path.
        let mut by_path: HashMap<String, pb::Reference> = HashMap::new();
        let mut order: Vec<String> = Vec::new();
        for rr in &r.references {
            let entry = by_path.entry(rr.path.clone()).or_insert_with(|| {
                order.push(rr.path.clone());
                pb::Reference {
                    path: rr.path.clone(),
                    target_kind: rr.target_kind.clone(),
                    target_name: rr.target_name.clone(),
                    unresolved: rr.unresolved.clone(),
                    object: Vec::new(),
                }
            });
            if let Some(obj) = &rr.object {
                for p in &obj.props {
                    if let Some(sv) = p.scalar() {
                        if let Some(c) = &sv.constraint {
                            entry.object.push(pb::PropertyConstraint {
                                path: p.path.clone(),
                                constraint: merge_constraints(std::slice::from_ref(c)),
                            });
                        }
                    }
                }
            }
        }
        order.sort();
        let references = order
            .into_iter()
            .map(|p| by_path.remove(&p).unwrap())
            .collect();

        pb::Resource {
            kind: r.resource.kind.clone(),
            name: r.resource.name.clone(),
            managed: self.is_managed(&r.resource.kind),
            properties,
            references,
        }
    }
}

/// Collects the scalar constraint expressions for a property path from every
/// rule that matched the resource.
fn prop_constraints(path: &str, matched: &[Rc<Rule>]) -> Vec<Constraint> {
    let mut cs = Vec::new();
    for r in matched {
        for p in &r.props {
            if p.path != path {
                continue;
            }
            if let Some(sv) = p.scalar() {
                if let Some(c) = &sv.constraint {
                    cs.push(c.clone());
                }
            }
        }
    }
    cs
}

/// Normalizes a set of constraints (all conjoined) into the wire Constraint:
/// required flag, min/max bounds, allowed/excluded value sets, plus the
/// effective constraint rendered into `expr`. Returns `None` if empty.
fn merge_constraints(cs: &[Constraint]) -> Option<pb::Constraint> {
    if cs.is_empty() {
        return None;
    }

    let mut expr_parts: Vec<String> = Vec::new();
    let mut seen: HashSet<String> = HashSet::new();
    let mut required = false;
    let mut lower: Option<Comparison> = None;
    let mut upper: Option<Comparison> = None;
    let mut exact: Option<Comparison> = None;
    let mut excluded: Vec<Value> = Vec::new();
    let mut sets: Vec<Vec<Value>> = Vec::new();

    for c in cs {
        let s = c.to_string();
        if seen.insert(s.clone()) {
            expr_parts.push(s);
        }
        let mut terms = Vec::new();
        conjuncts(c, &mut terms);
        for term in terms {
            match term {
                Constraint::Required(_) => required = true,
                Constraint::Comparison(cmp) => match cmp.op {
                    CompareOp::Eq => exact = Some(cmp),
                    CompareOp::Neq => excluded.push(cmp.value),
                    CompareOp::Ge | CompareOp::Gt => {
                        let tighter = match &lower {
                            None => true,
                            Some(l) => tighter_lower(&cmp, l),
                        };
                        if tighter {
                            lower = Some(cmp);
                        }
                    }
                    CompareOp::Le | CompareOp::Lt => {
                        let tighter = match &upper {
                            None => true,
                            Some(u) => tighter_upper(&cmp, u),
                        };
                        if tighter {
                            upper = Some(cmp);
                        }
                    }
                },
                Constraint::Or(alts) => {
                    if let Some(comps) = exact_alternatives(&alts) {
                        sets.push(comps.into_iter().map(|c| c.value).collect());
                    }
                }
                _ => {}
            }
        }
    }

    let mut out = pb::Constraint {
        required,
        expr: expr_parts.join(" & "),
        ..Default::default()
    };
    if let Some(l) = &lower {
        out.min = Some(pb::Bound {
            value: Some(value_to_proto(&l.value)),
            inclusive: l.op == CompareOp::Ge,
        });
    }
    if let Some(u) = &upper {
        out.max = Some(pb::Bound {
            value: Some(value_to_proto(&u.value)),
            inclusive: u.op == CompareOp::Le,
        });
    }

    let allowed: Vec<Value> = if let Some(e) = &exact {
        vec![e.value.clone()]
    } else if let Some((first, rest)) = sets.split_first() {
        let mut a = first.clone();
        for s in rest {
            a.retain(|v| value_in_set(v, s));
        }
        a
    } else {
        Vec::new()
    };
    out.allowed = allowed.iter().map(value_to_proto).collect();
    out.excluded = excluded.iter().map(value_to_proto).collect();
    Some(out)
}

fn value_to_proto(v: &Value) -> pb::Value {
    let (kind, unit) = match v.kind {
        ValueKind::Number => (pb::value::Kind::NumberValue(v.num), String::new()),
        ValueKind::Bool => (pb::value::Kind::BoolValue(v.bool), String::new()),
        ValueKind::String => (pb::value::Kind::StringValue(v.str.clone()), String::new()),
        ValueKind::Size => (pb::value::Kind::SizeBytes(v.num), v.unit.clone()),
        ValueKind::Duration => (pb::value::Kind::DurationMs(v.num), v.unit.clone()),
    };
    pb::Value {
        kind: Some(kind),
        unit,
    }
}

fn source_to_proto(s: ValueSource) -> i32 {
    match s {
        ValueSource::Explicit => pb::ValueSource::Explicit as i32,
        ValueSource::Default => pb::ValueSource::Default as i32,
    }
}

#[cfg(test)]
mod tests {
    use crate::eval::Resource;
    use crate::pb;
    use crate::testutil::{parse_set, str_attrs};

    fn resource<'a>(p: &'a pb::EvaluationResult, kind: &str, name: &str) -> &'a pb::Resource {
        p.resources
            .iter()
            .find(|r| r.kind == kind && r.name == name)
            .expect("resource")
    }
    fn property<'a>(r: &'a pb::Resource, path: &str) -> &'a pb::Property {
        r.properties
            .iter()
            .find(|p| p.path == path)
            .expect("property")
    }
    fn number(v: &Option<pb::Value>) -> f64 {
        match v.as_ref().and_then(|v| v.kind.as_ref()) {
            Some(pb::value::Kind::NumberValue(n)) => *n,
            other => panic!("not a number: {other:?}"),
        }
    }
    fn duration_ms(v: &Option<pb::Value>) -> f64 {
        match v.as_ref().and_then(|v| v.kind.as_ref()) {
            Some(pb::value::Kind::DurationMs(n)) => *n,
            other => panic!("not a duration: {other:?}"),
        }
    }
    fn string_val(v: &Option<pb::Value>) -> &str {
        match v.as_ref().and_then(|v| v.kind.as_ref()) {
            Some(pb::value::Kind::StringValue(s)) => s,
            other => panic!("not a string: {other:?}"),
        }
    }

    #[test]
    fn to_proto() {
        let rs = parse_set(
            "\nif env.type == \"production\" {\n    for service {\n        cpu: >= 1 & <= 4 | default 2\n    }\n    sql_cluster \"main\" {\n        engine: \"postgres\"\n        backup_retention: required & >= 30d | default 30d\n    }\n    for sql_database {\n        cluster: sql_cluster.main & {\n            backup_retention: >= 30d\n        }\n    }\n}\n",
        );
        let er = rs
            .evaluate_env(
                &str_attrs(&[("env.type", "production")]),
                &[
                    Resource {
                        kind: "service".into(),
                        name: "api".into(),
                        ..Default::default()
                    },
                    Resource {
                        kind: "sql_database".into(),
                        name: "orders".into(),
                        ..Default::default()
                    },
                ],
            )
            .unwrap();
        let p = rs.to_proto(&er);
        assert_eq!(p.resources.len(), 3);

        let day = (24 * 60 * 60 * 1000) as f64;

        // service "api": app-discovered; cpu defaulted to 2, constrained [1, 4].
        let api = resource(&p, "service", "api");
        assert!(!api.managed);
        let cpu = property(api, "cpu");
        assert_eq!(number(&cpu.value), 2.0);
        assert_eq!(cpu.source, pb::ValueSource::Default as i32);
        let cc = cpu.constraint.as_ref().unwrap();
        assert!(!cc.required);
        let min = cc.min.as_ref().unwrap();
        assert_eq!(number(&min.value), 1.0);
        assert!(min.inclusive);
        assert_eq!(number(&cc.max.as_ref().unwrap().value), 4.0);
        assert_eq!(cc.expr, ">= 1 & <= 4");

        // sql_cluster "main": managed and instantiated.
        let main = resource(&p, "sql_cluster", "main");
        assert!(main.managed);
        assert_eq!(string_val(&property(main, "engine").value), "postgres");
        let br = property(main, "backup_retention");
        assert_eq!(duration_ms(&br.value), 30.0 * day);
        assert_eq!(br.value.as_ref().unwrap().unit, "d");
        assert_eq!(br.source, pb::ValueSource::Default as i32);
        let bc = br.constraint.as_ref().unwrap();
        assert!(bc.required);
        assert_eq!(duration_ms(&bc.min.as_ref().unwrap().value), 30.0 * day);
        assert_eq!(bc.expr, "required & >= 30d");

        // sql_database "orders": cluster reference resolved to main + object
        // constraint.
        let orders = resource(&p, "sql_database", "orders");
        assert_eq!(orders.references.len(), 1);
        let r = &orders.references[0];
        assert_eq!(r.path, "cluster");
        assert_eq!(r.target_kind, "sql_cluster");
        assert_eq!(r.target_name, "main");
        assert_eq!(r.unresolved, "");
        assert_eq!(r.object.len(), 1);
        assert_eq!(r.object[0].path, "backup_retention");
        let oc = r.object[0].constraint.as_ref().unwrap();
        assert_eq!(duration_ms(&oc.min.as_ref().unwrap().value), 30.0 * day);
    }
}
